packages/puppeteer-core/src/bidi/Frame.ts

```diff
@@ -3,263 +3,701 @@
+
+import type {BidiBrowserContext} from './BrowserContext.js';
+import {BidiCdpSession} from './CdpSession.js';
+import type {BrowsingContext} from './core/BrowsingContext.js';
+import {BidiDeserializer} from './Deserializer.js';
+import {BidiDialog} from './Dialog.js';
+import {BidiElementHandle} from './ElementHandle.js';
 import {ExposeableFunction} from './ExposedFunction.js';
+import {BidiHTTPRequest, requests} from './HTTPRequest.js';
 import type {BidiHTTPResponse} from './HTTPResponse.js';
-import {
-  getBiDiLifecycleEvent,
-  getBiDiReadinessState,
-  rewriteNavigationError,
-} from './lifecycle.js';
+import {BidiJSHandle} from './JSHandle.js';
 import type {BidiPage} from './Page.js';
-import {
-  MAIN_SANDBOX,
-  PUPPETEER_SANDBOX,
-  Sandbox,
-  type SandboxChart,
-} from './Sandbox.js';
+import {BidiFrameRealm} from './Realm.js';
+
+const frames = new WeakMap<BrowsingContext, BidiFrame>();
 
 /**
  * Puppeteer's Frame class could be viewed as a BiDi BrowsingContext implementation
  * @internal
  */
 export class BidiFrame extends Frame {
-  #page: BidiPage;
-  #context: BrowsingContext;
-  #timeoutSettings: TimeoutSettings;
-  #abortDeferred = Deferred.create<never>();
-  #disposed = false;
-  sandboxes: SandboxChart;
-  override _id: string;
-
-  constructor(
+  static from(
+    context: BrowsingContext,
     page: BidiPage,
+    timeoutSettings: TimeoutSettings
+  ): BidiFrame {
+    const frame = new BidiFrame(context, page, timeoutSettings);
+    frame.#initialize();
+    return frame;
+  }
+
+  // keep-sorted start
+  // TODO: Replace with weak ref.
+  readonly #browsingContext: BrowsingContext;
+  readonly #page: BidiPage;
+  readonly #timeoutSettings: TimeoutSettings;
+  readonly destroyed$: Observable<never>;
+  override readonly client: BidiCdpSession;
+  readonly realms: {default: BidiFrameRealm; internal: BidiFrameRealm};
+  // keep-sorted end
+
+  private constructor(
     context: BrowsingContext,
-    timeoutSettings: TimeoutSettings,
-    parentId?: string | null
+    page: BidiPage,
+    timeoutSettings: TimeoutSettings
   ) {
     super();
+    frames.set(context, this);
+
+    this.#browsingContext = context;
     this.#page = page;
-    this.#context = context;
     this.#timeoutSettings = timeoutSettings;
-    this._id = this.#context.id;
-    this._parentId = parentId ?? undefined;
 
-    this.sandboxes = {
-      [MAIN_SANDBOX]: new Sandbox(undefined, this, context, timeoutSettings),
-      [PUPPETEER_SANDBOX]: new Sandbox(
-        UTILITY_WORLD_NAME,
+    this.client = new BidiCdpSession(this);
+    this.realms = {
+      default: new BidiFrameRealm(
+        this.#browsingContext.defaultRealm,
         this,
-        context.createRealmForSandbox(),
-        timeoutSettings
+        this.#timeoutSettings
+      ),
+      internal: new BidiFrameRealm(
+        this.#browsingContext.createWindowRealm(
+          `__puppeteer_internal_${Math.ceil(Math.random() * 10000)}`
```

Comment:
Why is this random, I think it adds low benefit while making it harder to debug?